pkg/util/hostname/providers.go

```diff
@@ -62,64 +64,74 @@ type provider struct {
 // * FQDN
 // * OS hostname
 // * EC2
-var providerCatalog = []provider{
-	{
-		name:             configProvider,
-		cb:               fromConfig,
-		stopIfSuccessful: true,
-		expvarName:       "'hostname' configuration/environment",
-	},
-	{
-		name:             "hostnameFile",
-		cb:               fromHostnameFile,
-		stopIfSuccessful: true,
-		expvarName:       "'hostname_file' configuration/environment",
-	},
-	{
-		name:             fargateProvider,
-		cb:               fromFargate,
-		stopIfSuccessful: true,
-		expvarName:       "fargate",
-	},
-	{
-		name:             "gce",
-		cb:               fromGCE,
-		stopIfSuccessful: true,
-		expvarName:       "gce",
-	},
-	{
-		name:             "azure",
-		cb:               fromAzure,
-		stopIfSuccessful: true,
-		expvarName:       "azure",
-	},
-
-	// The following providers are coupled. Their behavior changes depending on the result of the previous provider.
-	// Therefore 'stopIfSuccessful' is set to false.
-	{
-		name:             "fqdn",
-		cb:               fromFQDN,
-		stopIfSuccessful: false,
-		expvarName:       "fqdn",
-	},
-	{
-		name:             "container",
-		cb:               fromContainer,
-		stopIfSuccessful: false,
-		expvarName:       "container",
-	},
-	{
-		name:             "os",
-		cb:               fromOS,
-		stopIfSuccessful: false,
-		expvarName:       "os",
-	},
-	{
+func getProviderCatalog(notUseIMDSv2 bool) []provider {
+	providerCatalog := []provider{
```

Comment:
Would you mind keep the providers static in a var block and create with 2 independant AWS providers and then have the method choose which one to use ?

Something like:
```
var (
	configProvider = provider{
			name:             configProvider,
			cb:               fromConfig,
			stopIfSuccessful: true,
			expvarName:       "'hostname' configuration/environment",
	},
        [...]
)

func getProviderCatalog(withoutIMDSv2 bool) []provider {
	providers := []{
			configProvider,
			hostnameFileProvider,
			[...]
	}

	if withoutIMDSv2 (
		providers = append(providers, awsNoIMDSV2Provider)
	} else {
		providers = append(providers, awsIMDSV1Provider)
	}
	return providers 
}
```

Reply:
done